sdk/core/core-rest-pipeline/src/policies/multipartPolicy.ts

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { randomUUID, stringToUint8Array } from "@azure/core-util";
+import { BodyPart, HttpHeaders, MultipartRequestBody, RequestBodyType } from "../interfaces";
+import { PipelinePolicy } from "../pipeline";
+import { concatenateStreams, toStream } from "../util/stream";
+
+export const multipartPolicyName = "multipartPolicy";
+
+function generateBoundary(): string {
+  return `----AzSDKFormBoundary${randomUUID()}`;
+}
+
+function encodeHeaders(headers: HttpHeaders): string {
+  return [...headers].map(([name, value]) => `${name}: ${value}\r\n`).join("");
+}
+
+function createBodyStream(
+  parts: BodyPart[],
+  boundary: string
+): ReadableStream | NodeJS.ReadableStream {
+  const streams = [
+    stringToUint8Array(`--${boundary}`, "utf-8"),
+    ...parts.flatMap((part) => [
+      stringToUint8Array("\r\n", "utf-8"),
+      stringToUint8Array(encodeHeaders(part.headers), "utf-8"),
+      stringToUint8Array("\r\n", "utf-8"),
+      part.body,
+      stringToUint8Array(`\r\n--${boundary}`, "utf-8"),
+    ]),
+    stringToUint8Array("--", "utf-8"),
+  ].map(toStream);
+
+  return concatenateStreams(streams);
+}
+
+export function isMultipartRequestBody(
+  body: RequestBodyType | undefined
+): body is MultipartRequestBody {
+  return Boolean(body && Array.isArray((body as MultipartRequestBody).parts));
+}
+
+export function multipartPolicy(): PipelinePolicy {
+  return {
+    name: multipartPolicyName,
+    sendRequest(request, next) {
+      if (!isMultipartRequestBody(request.body)) {
+        return next(request);
+      }
+
+      let boundary = request.body.boundary;
+
+      const contentTypeHeader = request.headers.get("Content-Type");
+      if (contentTypeHeader) {
+        const parsedHeader = contentTypeHeader.match(/^(multipart\/[^ ;]+)(; *boundary=(.+))?$/);
+        if (!parsedHeader) {
+          throw new Error(
+            `Got multipart request body, but content-type header was not multipart: ${contentTypeHeader}`
+          );
+        }
+
+        const parsedContentType = parsedHeader[1];
+        const parsedBoundary = parsedHeader[3];
```

Comment:
really silly nit: could we make the second capture group non-capturing so this can be offsets 1 and 2?